### PR TITLE
add user search to invite dialog

### PR DESCRIPTION
**Pull request by maltee1**
Please have a look, it probably needs a lot of improvements. Among other things, I hijacked the UsersModel, not sure if you're happy with that. What I'm still planning to do:
- improve local search to include more, possibly all, rooms (not in this PR)
- add the search to CreateDirect.qml (not in this PR)
- add a narrow mode since it should work on mobile

Are there better ways of displaying the result for direct mxid entry? it currently needs its own code block, would be nicer if it could work with the same code, but that would require a new model or changes to the existing ones.

### CODE REVIEW

File: src/UserDirectoryModel.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+void
+UserDirectoryModel::displaySearchResults(std::vector<mtx::responses::User> results)
+{
+    results_ = results;
+    if (results_.empty()) {
+        nhlog::net()->error("mtxclient helper thread yielded empty chunk!");
+        return;
+    }
+    beginInsertRows(QModelIndex(), 0, static_cast<int>(results_.size()) - 1);
```
Comment (deepbluev7) [inline src/UserDirectoryModel.cpp:85]: You need to call this before starting to do any modifications, otherwise the UI might break.

Comment (maltee1) [reply src/UserDirectoryModel.cpp:84]: like that?

**Comment by maltee1** [comment]
The narrow layout will be activated when the width of the window is less than 500 px. I stole the design from Element Android. I will probably add avatars in the future, but trying to do so messed up the layout and I don't have patience for that right now.
The row for the invitees list and search result list should probably get its own class, it's also used for the member list. Unfortunately, qml doesn't give me any helpful debugging messages, it just refuses to open the window so I can't really check what the problem is. I might also do that in the future.

**Comment by deepbluev7** [comment]
Usually we use these switch button thingies instead of checkboxes (check the roomsettings for example). Also the colors look off:

grafik


(And I get like no results, did this actually work for you?)

File: resources/qml/dialogs/InviteDialog.qml [hunk 2]
```diff
@@ -83,85 +121,99 @@ ApplicationWindow {
...
+                    id: searchTimer
+
+                    interval: 350
+                    onTriggered: {
+                        userSearch.model.setSearchString(parent.text)
+                    }
+                }
             }
 
-            Button {
```
Comment (deepbluev7) [inline resources/qml/dialogs/InviteDialog.qml:88]: Why did you remove this button? :3

File: src/InviteesModel.cpp [hunk 3]
```diff
@@ -85,21 +85,30 @@ InviteesModel::mxids()
...
-          if (err) {
-              nhlog::net()->warn("failed to retrieve profile info");
-              emit userInfoLoaded();
-              return;
-          }
-
-          displayName_ = QString::fromStdString(res.display_name);
-          avatarUrl_   = QString::fromStdString(res.avatar_url);
+    // checking for empty avatarUrl will cause profiles to be loaded that don't have an avatar
+    // to needlessly be loaded. Can we make sure we either provide both or none?
```
Comment (deepbluev7) [inline src/InviteesModel.cpp:93]: Do we need to? We can just always load it, can't we?

File: src/UsersModel.cpp [hunk 4]
```diff
@@ -9,17 +9,37 @@
...
+                                                          QString::fromStdString(userId)));
+                }
+            }
+        }
+    } else {
+        for (const auto &m : cache::roomMembers(roomId)) {
+            displayNames.push_back(QString::fromStdString(cache::displayName(room_id, m)));
+            userids.push_back(QString::fromStdString(m));
+            avatarUrls.push_back(
+              cache::avatarUrl(QString::fromStdString(room_id), QString::fromStdString(m)));
```
Comment (deepbluev7) [inline src/UsersModel.cpp:41]: This actually adds quite some cost in some rooms. It is fine for now, but remember me to optimize this, after this is merged :3

File: src/UserDirectoryModel.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+          emit searchingUsersChanged();
+
+          if (err) {
+              nhlog::net()->error("Failed to retrieve users from mtxclient - {} - {} - {}",
+                                  mtx::errors::to_string(err->matrix_error.errcode),
+                                  err->matrix_error.error,
+                                  err->parse_error);
+          } else {
+              emit fetchedSearchResults(res.results);
+          }
```
Comment (deepbluev7) [inline src/UserDirectoryModel.cpp:57]: This is actually unsafe. I noticed that only just now, but you might want to do something like this: https://github.com/Nheko-Reborn/nheko/commit/0b8709a0ead4bdc71666f09860a058c90e3c7a1a

File: resources/qml/dialogs/InviteDialog.qml (hunk 2 above)
Comment (maltee1) [reply resources/qml/dialogs/InviteDialog.qml:88]: Users are now added by clicking on them, not by clicking on the "add" button. It makes more sense if there's more than one user to choose from.

apparently I introduced the bug when I tried to call beginInsertRows() before making changes... fixed in the upcoming commit.

File: src/UsersModel.cpp (hunk 4 above)
Comment (maltee1) [reply src/UsersModel.cpp:41]: I was thinking this might be an issue, but I also thought that not using the avatarUrl for the friends "room" is adding unnecessary cache accesses. I didn't want to add additional logic for distinguishing the two and since the same process is used to get the displayNames, I thought the performance his is probably acceptable.

File: src/UserDirectoryModel.cpp (hunk 5 above)
Comment (maltee1) [reply src/UserDirectoryModel.cpp:67]: Not sure what you mean, is this OK?

File: resources/qml/dialogs/InviteDialog.qml (hunk 2 above)
Comment (deepbluev7) [reply resources/qml/dialogs/InviteDialog.qml:88]: Ah, I guess that makes sense. It wasn't that obvious to me, since I am used to the old design though, but maybe it is actually more intuitive.

File: src/UserDirectoryModel.cpp (hunk 1 above)
Comment (deepbluev7) [reply src/UserDirectoryModel.cpp:95]: Well, kinda, there is still a bug in there. Hint: You had the same bug 4 lines above.

**Comment by maltee1** [comment]
Separator added.

I currently cannot figure out why the user directory isn't working again. Shouldn't endInsertRows(); trigger a content update? endInsertRows() is executed after a search is submitted, but the QML never tries to fetch the data.

**Comment by deepbluev7** [comment]
endInsertRows() only tells the UI to render the new rows you just added. It only start fetching more, if it canFetchMore and you scroll to the bottom?

**Comment by maltee1** [comment]
Well, every time I call endInsertRows(), all rows will be newly added rows. Shouldn't that suffice? I don't entirely understand how canFetchMore() is used, but don't we set that to true to trigger fetching users from the server? Once we fetched, we would set it to false?

**Comment by deepbluev7** [comment]
Qt calls canFetchMore() once, when the model is loaded. If it isn't true at that time, it will never call fetchMore, which will never trigger any endInsertRows(), so there will never be anything loaded.

**Comment by maltee1** [comment]
So the one thing that seems to make it work is calling `results_=results` before `beginInsertRows()`. It did work at least once before I made that change, but apparently that's not reproducible anymore. You said that this might break the UI and also RoomDirectoryModel does not need that.
I have no idea why and how to proceed. But I *can* make it work ;)

**Comment by maltee1** [comment]
Working now.... when merging, could you squash the commits or let me know so I can squash them? this is actually a bit embarrassing :D

**Comment by deepbluev7** [comment]
> let me know so I can squash them

Squash them as you want. I will try to remember to squash merge, but I forget stuff like that all the time :D

**Comment by deepbluev7** [comment]
Small nitpick:

grafik

Maybe put the mxid there, if the displayname is empty :D

File: resources/qml/components/UserListRow.qml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+import ".."
+import QtQuick 2.12
+import QtQuick.Controls 2.12
+import QtQuick.Layouts 1.12
+import im.nheko 1.0
+
+ItemDelegate {
```
Comment (deepbluev7) [inline resources/qml/components/UserListRow.qml:13]: this looks very similar to the AvatarListTile in resources/qml/components

File: src/InviteesModel.cpp (hunk 3 above)
Comment (deepbluev7) [reply src/InviteesModel.cpp:93]: I guess that is fine, yes. Note that some users have an avatar, but no displayname. Maybe you want to pass the mxid in that case or check also for the avatarUrl here?

File: src/UserDirectoryModel.cpp (hunk 5 above)
Comment (deepbluev7) [reply src/UserDirectoryModel.cpp:70]: No, you are still not using the proxy as I did in that commit :)

File: src/UsersModel.cpp [hunk 7]
```diff
@@ -59,8 +79,7 @@ UsersModel::data(const QModelIndex &index, int role) const
         case CompletionModel::SearchRole2:
```
Comment (deepbluev7) [inline src/UsersModel.cpp:79]: The line above this should default to returning the mxid, if the displayname is not there. Otherwise some things have no name in some cases. Wasn't an issue before, but with your pills it is.

File: src/UserDirectoryModel.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+QVariant
+UserDirectoryModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid() || index.row() >= (int)results_.size() || index.row() < 0)
+        return {};
+    switch (role) {
+    case Roles::DisplayName:
+        return QString::fromStdString(results_[index.row()].display_name);
```
Comment (deepbluev7) [inline src/UserDirectoryModel.cpp:82]: Here falling back to the mxid also seems sensible.

File: resources/qml/components/UserListRow.qml (hunk 6 above)
Comment (maltee1) [reply resources/qml/components/UserListRow.qml:13]: I didn't see that one because only the imagepack thing uses it -.- Are you sure it's suitable in its current form, and, if not, does it make sense to make it suitable? One problem I see is that I'd be dependent on buggy tap handlers to handle clicks, but I guess I could fix that. I'm not sure what other problems there might be. It just seems like my UserListRow is simpler, albeit possibly less versatile.

File: src/UsersModel.cpp (hunk 7 above)
Comment (maltee1) [reply src/UsersModel.cpp:79]: I see the issue, but I'd rather fix that in the QML. Returning an mxid as a display name seems wrong because they are not the same. Also, showing a displayname only when a profile is available is a good indicator that the mxid was entered without typo. Checking whether mxid and displayname are the same is less intuitive and also slower.

File: src/UserDirectoryModel.cpp (hunk 8 above)
Comment (maltee1) [reply src/UserDirectoryModel.cpp:82]: Again, I'd rather not, because an mxid is not a display name. Changing the QML will actually fix both cases ;)

File: src/InviteesModel.cpp (hunk 3 above)
Comment (maltee1) [reply src/InviteesModel.cpp:93]: I didn't consider the case of an empty displayname. Possibly rather than passing empty strings, we need to specify whether user info should be loaded or not, since we know that. I just decided to use what's there and not change function declarations.

Is it OK to assume that if both avatarUrl and displayName are empty, we need to fetch a profile? If only one is empty, we can assume that profile info is complete, if both are empty we may either not have passed them and thus need to fetch the profile or the user has indeed empty displayname and avatarUrl, but fetching the profile isn't going to break anything, it's just an unnecessary request.

File: src/UserDirectoryModel.cpp (hunk 5 above)
Comment (maltee1) [reply src/UserDirectoryModel.cpp]: I have no idea what I am doing, but does this work?

File: src/UsersModel.cpp (hunk 7 above)
Comment (deepbluev7) [reply src/UsersModel.cpp:79]: Yeah, I guess that works too.

File: resources/qml/components/UserListRow.qml (hunk 6 above)
Comment (deepbluev7) [reply resources/qml/components/UserListRow.qml:13]: I don't really care tbh. I just wanted to mention it, since you probably hadn't seen it.

**Comment by deepbluev7** [comment]
Seems to work fine now, thanks for bearing with me!
